### PR TITLE
Make surgery a player_activity

### LINKED ISSUE #17698: Tinkering with bionics should cost time
A lot of time! As now both installing and uninstalling them is completely time-free.

### PR DESCRIPTION
**Pull request by Fris0uman**
#### Summary
```SUMMARY: Balance "Makes surgery a player_activity"```

#### Purpose of change

Surgery takes time instead of instantly adding/removing a cbm.

#### Describe the solution

 - [x] Move uninstall to an activity
 - [x] Move install to an activity
 - [x] Make damage of surgery body part specific

#### Describe alternatives you've considered

Just checking for an under_operation effect, if the character currently on the couch has it then the autodoc can't be used again until the end of the effect.
#### Additional context

- The operation happens in three phases:
    - First for half the duration the Autodoc opens the patient
    - The installation/removal happens
    - For the second half of duration the Autodoc stiches the patient back up or attempts to if the operation was a failure.

 - You can't start another operation before the first one is finished
 - Operations can't be interrupted
 - Patient are strapped in the autodoc and can't be dragged away by grapplers
 - Destroying the autodoc or the autodoc couch during the opration (Or debug teleporting the patien away) will result in: 
                             -  Bleeding on all body part involved for a duration of `infinite`and an intensity equal to surgery `Difficulty`
                              -  `20 * Difficulty` dammage on all body part involved so damage ranging between 0 and 280
 - Failure message is always `"The installation is a failure."`or `"The removal is a failure."`, instead of the 1 among 5 random messages meaning the same thing we had before, followed by description of the consequences. This is because player_activity don't store player pointers.
- fixes #17698 : npc surgeon is now slightly less magical

Overall this should have minor balance consequences and allow to eventually expend surgery either with more detailed flavour about what's going on or with actual game mechanic.

### CODE REVIEW

File: src/iexamine.cpp [hunk 1]
```diff
@@ -4093,6 +4094,11 @@ void iexamine::autodoc( player &p, const tripoint &examp )
             popup( _( "No patient found located on the connected couches.  Operation impossible.  Exiting." ) );
             return;
         }
+    } else if( patient.has_effect( effect_under_op ) ) {
+        popup( _( "Operation underway.  Please wait the end of the current procedure.  Estimated time remaining: %s." ),
```
Comment (matskuman5) [inline src/iexamine.cpp]:
```suggestion
        popup( _( "Operation underway.  Please wait until the end of the current procedure.  Estimated time remaining: %s." ),
```

File: src/iexamine.cpp [hunk 2]
```diff
@@ -4093,6 +4094,11 @@ void iexamine::autodoc( player &p, const tripoint &examp )
...
+               to_string( patient.get_effect_dur( effect_under_op ) ) );
+        p.add_msg_if_player( m_info, _( "The autodoc is working on %s" ), patient.disp_name() );
```
Comment (matskuman5) [inline src/iexamine.cpp]:
```suggestion
        p.add_msg_if_player( m_info, _( "The autodoc is working on %s." ), patient.disp_name() );
```

File: src/player_activity.h [hunk 3]
```diff
@@ -40,6 +40,7 @@ class player_activity
         std::string name;
         std::vector<item_location> targets;
         std::vector<int> values;
+        std::vector<float> f_values;
```
Comment (ifreund) [inline src/player_activity.h]: If we're going to add floating point values to `player_activity`, I think we should use `double`.

However in this case it looks like you don't really need to store the value and can just recalculate it when you need it.

Comment (Fris0uman) [reply src/player_activity.h]: It's removed

File: src/activity_handlers.cpp [hunk 4]
```diff
@@ -2777,6 +3072,54 @@ void activity_handlers::try_sleep_finish( player_activity *act, player *p )
...
+                       _( "a short upbeat jingle: \"Operation Succesfull\"" ), true,
+                       "Autodoc",
+                       "success" );
+    } else {
+        add_msg( m_bad,
+                 _( "The Autodoc jerks back to its resting position after failing the removal." ) );
+        const std::list<tripoint> autodocs = g->m.find_furnitures_in_radius( p->pos(), 1,
+                                             furn_str_id( "f_autodoc" ) );
+        sounds::sound( autodocs.front(), 10, sounds::sound_t::music,
+                       _( "a sad beeping noise: \"Operation Failed\"" ), true,
```
Comment (Night-Pryanik) [inline src/activity_handlers.cpp]:
```suggestion
                       _( "a sad beeping noise: \"Operation failed\"" ), true,
```

File: src/activity_handlers.cpp [hunk 5]
```diff
@@ -2777,6 +3072,54 @@ void activity_handlers::try_sleep_finish( player_activity *act, player *p )
...
+                       _( "a short upbeat jingle: \"Operation Succesfull\"" ), true,
+                       "Autodoc",
+                       "success" );
+    } else {
+        add_msg( m_bad,
+                 _( "The Autodoc jerks back to its resting position after failing the installation." ) );
+        const std::list<tripoint> autodocs = g->m.find_furnitures_in_radius( p->pos(), 1,
+                                             furn_str_id( "f_autodoc" ) );
+        sounds::sound( autodocs.front(), 10, sounds::sound_t::music,
+                       _( "a sad beeping noise: \"Operation Failed\"" ), true,
```
Comment (Night-Pryanik) [inline src/activity_handlers.cpp]:
```suggestion
                       _( "a sad beeping noise: \"Operation failed\"" ), true,
```

File: src/activity_handlers.cpp [hunk 6]
```diff
@@ -2769,6 +2777,293 @@ void activity_handlers::try_sleep_do_turn( player_activity *act, player *p )
...
+        } else {
+            p->add_effect( effect_bleed, 1_turns, num_bp, true, difficulty );
+            p->apply_damage( nullptr, num_bp, 10 * difficulty );
+        }
+    }
+
+    if( time_left >  op_duration ) {
+        if( act->values.size() > 4 ) {
+            for( size_t i = 4; i < act->values.size(); i ++ ) {
+                if( calendar::once_every( 2_minutes ) && u_see ) {
```
[the remark this answers is not in the document]
Comment (Fris0uman) [reply src/activity_handlers.cpp]: Is 5 minutes better?

[the remark this answers is not in the document]
Comment (Night-Pryanik) [reply src/activity_handlers.cpp]: Yeah. If even 5 minutes would look like spam, we could further lower it in other PRs.

File: src/activity_handlers.cpp [hunk 7]
```diff
@@ -2769,6 +2775,146 @@ void activity_handlers::try_sleep_do_turn( player_activity *act, player *p )
...
 }
 
+void activity_handlers::operation_do_turn( player_activity *act, player *p )
+{
+    /**
+    - values[0]: Difficulty
+    - values[1]: success
+    - values[2]: max_power_level
+    - values[3]: pl_skill
+    - values[4: ]: occupied body_parts
```
Comment (Night-Pryanik) [inline src/activity_handlers.cpp]:
```suggestion
    - values[4]: occupied body_parts
```

File: src/activity_handlers.cpp [hunk 8]
```diff
@@ -2769,6 +2775,146 @@ void activity_handlers::try_sleep_do_turn( player_activity *act, player *p )
...
+    - str_values[0]: cbm name
+    - str_values[1]: bionic_id
+    - str_values[2]: upgraded cbm name
+    - str_values[3]: upgraded cbm bionic_id
+    - str_values[4]: installer name
+    - str_values[5: ]: traits removed by the cbm
```
Comment (Night-Pryanik) [inline src/activity_handlers.cpp]:
```suggestion
    - str_values[5]: traits removed by the cbm
```

File: src/activity_handlers.cpp (hunk 7 above)
Comment (Fris0uman) [reply src/activity_handlers.cpp]: So I know it's not the right synthaxe but I mean starting at 4 and above

File: src/activity_handlers.cpp (hunk 8 above)
Comment (Fris0uman) [reply src/activity_handlers.cpp]: Same here 5 and above

**Comment by KorGgenT** [comment]
jenkins rebuild

**Comment by AMurkin** [comment]
How is installation different from uninstallation inside the `activity_handlers::operation_do_turn`?
And why `activity.str_values` have only 2 elements inside `player::uninstall_bionic`, but `activity_handlers::operation_do_turn` checks `act->str_values[5]`?

**Pull request merged**